Add support for metadata API

mateiidavid opened the pull request:
## Motivation

In order to reduce memory and I/O pressure, we can use the (undocumented) [metadata API](https://github.com/kubernetes/client-go/blob/be20b2bd91d8997dd6e42647be047973acb8cdee/metadata/metadata.go#L180-L306). Instead of retrieving an entire object, we can retrieve only the metadata. This is useful in lots of scenarios where users might not be interested in the `spec` or `status`.


<!--
Explain the context and why you're making that change. What is the problem you're trying to solve?
If a new feature is being added, describe the intended use case that feature fulfills.
-->

## Solution

Metadata-only requests are made possible through the REST API by including an extended Accept header. The header will notify the API Server a JSON response is expected, and it will encode an additional 'as=PartialObjectMetadata' k=v pair to receive only the metadata of the object. This is more efficient when only the metadata needs to be parsed (as opposed to the whole spec).

To make it possible to use the API, this change adds new (metadata specific) request impls in `kube-core`, and new metadata methods on the `Api<K>` type. Unfortunately, the methods on the Api type get complicated really quickly when attempting to preserve the function signatures originally posted in #1027. Deserializing the wire format we have in the response directly into an `ObjectMeta` type works if we first deserialize into a `Value` and then into the concrete type. For wrapper types (such as in `list_metadata` or `watch_metadata`) this gets even more complex.

To get around it, I've added a new `PartialObjectMeta` type that is supposed to mirror the Go counterpart. A wrapper list has also been added. Both of these support conversions to `ObjectMeta` (or `ObjectList<ObjectMeta>`). This makes it a bit neater to implement everything.

This is still WIP (so the code is not super tidy), I wanted to show the progress and find a consensus moving forward (or even some ideas on how we could deserialize this better if we're not happy with the `PartialObject` types).

<details>

<summary> Sending requests through curl  </summary>

Thought it might be helpful. This helped me look at the format a bit more closely.

First, run `kubectl proxy &` and then send requests with curl

```
# Get a deployment's metadata
curl -H "Accept: application/json;as=PartialObjectMetadata;g=meta.k8s.io;v=v1" http://localhost:8001/apis/apps/v1/namespaces/kube-system/d
eployments/foo

# List a deployment(s) metadata
 curl -H "Accept: application/json;as=PartialObjectMetadataList;g=meta.k8s.io;v=v1" http://localhost:8001/apis/apps/v1/namespaces/kube-syst
em/deployments

# Watch
curl -H "Accept: application/json;as=PartialObjectMetadata;g=meta.k8s.io;v=v1" "http://localhost:8001/api/v1/namespaces/kube-system/pods?&
watch=true&resourceVersion=0&timeoutSeconds=290&allowWatchBookmarks=true"
```

</details>


Closes https://github.com/kube-rs/kube/issues/1027

Signed-off-by: Matei David <matei@buoyant.io>

clux commented on kube-client/src/api/core_methods.rs:
```diff
@@ -35,6 +41,34 @@ where
+            .request
+            .watch_metadata(lp, version)
+            .map_err(Error::BuildRequest)?;
+        req.extensions_mut().insert("watch");
+        self.client.request_events::<PartialObjectMeta>(req).await
```
if this works, then this is perfect; a mirror of what the api does (what the crate is meant to do).


..it will present some extra logic to `watcher` to actually set up a persistent watch stream (watcher assumes the input type is the same as the output type, and now we need to go through conversion first), but hopefully we can do something similar to `watch_object` and specialize for that down the line (maybe some `Into<K>` bounds can do it).

mateiidavid replied:
By persistent here we mean watches that do not terminate on errors? I have to be honest, haven't spent a lot of time looking and understanding how they're implemented in `kube-rs`. I looked at `watch_object`, think it might be work with some bounds. Happy to take a stab at it once this wraps up. Will try to do it before a release is out 🤞🏻

clux replied:
Yeah, most informer-style(watcher)/reflector-style/controller based workloads will want to use an infinite watch that retries on de-syncs. Regular `Api::watch` is basically awful for real-world use-cases because of how temperamental the api is, so we forward people to `watcher` or the more specific `watch_object`.

> Happy to take a stab at it once this wraps up. Will try to do it before a release is out 🤞🏻

That would be great ❤️

clux commented on kube-client/src/lib.rs:
```diff
@@ -470,6 +469,66 @@ mod test {
+                .labels
+                .expect("Expected metadata to have a 'name' field")
+                .get_key_value("app")
+        );
```
test is great. maybe also do a patch-metadata in this fn as well!
watch stuff we can relegate to testing in runtime.

mateiidavid replied:
Yeah, great call! For completeness I added a `list_metadata` test too. I was a bit apprehensive at first because there may be lots of pods running at the same time in the namespace. I managed to make it look nice through the `find` combinator.

clux replied:
It should be fine to do a list 👍 . Integration tests are usually set up in a minimal environment on CI with just a few pods, and we warn in most places that integration tests hits your current context and may mutate.

clux reviewed (commented):
This is great! Don't worry that the type signatures doesn't match what we originally set out. Matching the api is the actual goal. Left some minor comments on testing and formatting, but beyond that I think it's basically only docs that's needed here.

Watcher extensions in runtime is also an essential down the line, but we can do this here first.

mateiidavid commented:
Thanks a lot for the review! Added docs, might need some wordsmithing though. Thanks for the nightly toolchain and just recipe tip (yay for more just! makes my life easier). Fixed a unit test, put up docs, expanded the integration test. Let's hope we're all green :D

clux commented on kube-core/src/metadata.rs:
```diff
@@ -12,3 +14,50 @@ pub struct TypeMeta {
+/// Represents a list of objects and contains only the ObjectMeta
+#[derive(Deserialize, Serialize, Clone, Default, Debug)]
+pub struct PartialObjectMetaList {
+    /// The type fields
+    #[serde(flatten, default)]
+    pub types: Option<TypeMeta>,
```
AFAIKT this is the inclusion of this type (that we do not use) is the only difference between `PartialObjectMetaList<PartialObjectMeta>` and `ObjectList`. Given there are no other bounds constraints on `ObjectList` i'm leaning towards just using that unless there's more info in the partial that __could__ be useful.

That would avoid the (largely pointless-feeling) converter between them.

clux replied:
While upstream does have a specific List type for this as well, it follows [their regular list type conventions](https://github.com/kubernetes/api/blob/0478a3e95231398d8b380dc2a1905972be8ae1d5/apps/v1/types.go#L551-L560), so we should re-use `ObjectList<T>`. We have used `ObjectList` as the generic representation of that struct for ages and it continues to fit here with no extra fields.

mateiidavid replied:
Cool! I made the change. Didn't really do an impl for `From<ObjectList<PartialObjectMeta>> for ObjectList<PartialObjectMeta>` since I'm not sure we need it? Users can do a map or access the fields directly. If it makes sense to have as a util when we do the watcher I might add it back. Can also do that as part of this change if you think it's a good idea.

clux replied:
> `From<ObjectList<PartialObjectMeta>> for ObjectList<PartialObjectMeta`

uh, do you mean:
`From<ObjectList<PartialObjectMeta>> for ObjectList<ObjectMeta>` ?

if so, then nah, this is fine.

mateiidavid replied:
Yeah, oops meant `for ObjectList<ObjectMeta>`. All the objects and the metas got to me. But cool, makes sense 👍🏻

clux commented on kube-core/src/metadata.rs:
```diff
@@ -12,3 +14,50 @@ pub struct TypeMeta {
+/// Generic representation of any object with ObjectMeta. Allows clients to get
+/// access to a particular ObjectMeta schema without knowing the details of the
+/// version
+#[derive(Deserialize, Serialize, Clone, Default, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct PartialObjectMeta {
```
Given that this is actually a wrapper container around a object, I think we should maybe call this `PartialObject`

clux replied:
actually, no. let's defer to apimachinery for the root type: https://github.com/kubernetes/apimachinery/blob/7687996c715ee7d5c8cf1e3215e607eb065a4221/pkg/apis/meta/v1/types.go#L1416-L1425 (`PartialObject` implies that is could be deconstructed in other ways, and our `DynamicObject` type is kind of called Unstructured upstream)

oh, and btw, you can even take docstrings from there.

mateiidavid replied:
Mm yeah that makes sense. I also updated with client-go's docstrings 👍🏻

clux commented on kube-client/src/lib.rs:
```diff
@@ -469,6 +469,104 @@ mod test {
+        let pod_metadata = p_list
+            .items
+            .into_iter()
+            .find(|p| p.metadata.name.as_ref().expect("Pod must have a name") == "busybox-kube-meta")
```
```suggestion
            .find(|p| p.name_any() == "busybox-kube-meta")
```
ought to work.. maybe we need to `impl Resource` for `PartialObjectMeta`

clux replied:
you can probably just copy-paste the `impl Resource for Object` setup in `object.rs`

clux replied:
this will also allow you to shorten a bunch of your asserts above using`ResourceExt` - but it's also probably the magic sauce that will make `watcher` work for it :-)

clux commented on kube-core/src/metadata.rs:
```diff
@@ -12,3 +12,23 @@ pub struct TypeMeta {
     pub kind: String,
 }
+
+/// PartialObjectMetadata is a generic representation of any object with
+/// ObjectMeta. It allows clients to get access to a particular ObjectMeta
+/// schema without knowing the details of the version.
```
```suggestion
/// A generic representation of any object with `ObjectMeta`.
///
/// It allows clients to get access to a particular `ObjectMeta`
/// schema without knowing the details of the version.
```
slight reformat for rustdoc

clux reviewed (approved):
This is all sensible to me. As discussed on discord, a watcher setup would be nice, but it's non-trivial without any further indirection so let's put a clear break here for now.

CI is green apart from the unrelated deny issue, so all good for me. Can merge tomorrow unless you have objections.

mateiidavid commented:
> This is all sensible to me. As discussed on discord, a watcher setup would be nice, but it's non-trivial without any further indirection so let's put a clear break here for now.
> 
> 
> 
> CI is green (EDIT: github was lagging, hopefully still after the approval) apart from the unrelated deny issue, so all good for me. Can merge tomorrow unless you have objections.


No objections :) will start playing with the code to see how we can get a watcher setup working
